File: VindsidenApp/RHCAppDelegate.m
//
//  RHCAppDelegate.m
//  Vindsiden-v2
//
//  Created by Ragnar Henriksen on 01.05.13.
//  Copyright (c) 2013 RHC. All rights reserved.
//

#import <AFNetworking/AFNetworkActivityIndicatorManager.h>
#import "RHCAppDelegate.h"
#import "RHCViewController.h"
#import "RHEVindsidenAPIClient.h"

@import VindsidenKit;

@interface RHCAppDelegate ()

@end


@implementation RHCAppDelegate
{
    dispatch_queue_t _formatterQueue;
}


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    [[AFNetworkActivityIndicatorManager sharedManager] setEnabled:YES];

    self.window.tintColor = RGBCOLOR( 227.0, 60.0, 13.0);

    _formatterQueue = dispatch_queue_create("formatter queue", NULL);
    
////#ifndef DEBUG
//    [TestFlight setOptions:@{@"disableInAppUpdates":@NO}];
//    [TestFlight setDeviceIdentifier:[[UIDevice currentDevice] uniqueIdentifier]];
//
////#endif
//    [TestFlight takeOff:@"b97e9c55-3aae-4f18-bc27-6fac0e973438"];

    // Override point for customization after application launch.
    if ( [[Datamanager sharedManager].sharedDefaults integerForKey:@"selectedUnit"] == 0 ) {
        [[Datamanager sharedManager].sharedDefaults setInteger:SpeedConvertionToMetersPerSecond forKey:@"selectedUnit"];
        [[Datamanager sharedManager].sharedDefaults synchronize];

    }

    [[Datamanager sharedManager] cleanupPlots];

    [application setMinimumBackgroundFetchInterval:UIApplicationBackgroundFetchIntervalMinimum];
    return YES;
}


- (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    NSURL *url = launchOptions[UIApplicationLaunchOptionsURLKey];
    if ( url.host == nil || [url.host rangeOfString:@"station" options:NSCaseInsensitiveSearch].location == NSNotFound ) {
        return NO;
    }
    return YES;
}


- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation
{
    DLOG(@"url: %@", url);

    if ( url.host == nil || [url.host rangeOfString:@"station" options:NSCaseInsensitiveSearch].location == NSNotFound ) {
        return NO;
    }

    return [self openLaunchOptionsURL:url];
}


- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.

    [RHEVindsidenAPIClient defaultManager].background = YES;
}


- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}


- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}


- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    [RHEVindsidenAPIClient defaultManager].background = NO;
}


- (void)applicationWillTerminate:(UIApplication *)application
{
    [[Datamanager sharedManager] saveContext];
}


#if 0
- (void)applicationProtectedDataDidBecomeAvailable:(UIApplication *)application
{
    DLOG(@"");
}


- (void)applicationProtectedDataWillBecomeUnavailable:(UIApplication *)application
{
    DLOG(@"");
}
#endif


- (void)application:(UIApplication *)application performFetchWithCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
{
    DLOG(@"fetch");
    UINavigationController *nc = (UINavigationController *)self.window.rootViewController;
    RHCViewController *vc = (RHCViewController *)[nc.viewControllers firstObject];

    [vc updateContentWithCompletionHandler:completionHandler];
}


#pragma mark - Application's Documents directory


- (NSString *) applicationPrivateDocumentsDirectory
{
    NSFileManager *fm = [NSFileManager defaultManager];
    NSString *path = [[NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES) lastObject] stringByAppendingPathComponent:@"Private Documents"];

    [fm createDirectoryAtPath:path withIntermediateDirectories:YES
                   attributes:nil
                        error:nil];

    NSError *error = nil;
    NSURL *url = [NSURL fileURLWithPath:path];
    [url setResourceValue:@YES forKey:NSURLIsExcludedFromBackupKey error:&error];
    return path;
}


#pragma mark - 


- (BOOL)openLaunchOptionsURL:(NSURL *)url
{
    id ident = [url.pathComponents lastObject];
    CDStation *station = nil;
    if ( [ident isNumeric] ) {
        station = [CDStation existingStation:ident inManagedObjectContext:[[Datamanager sharedManager] managedObjectContext]];
    } else {
        station = [CDStation searchForStation:ident inManagedObjectContext:[[Datamanager sharedManager] managedObjectContext]];
    }

    if ( nil == station ) {
        return NO;
    }

    if ( [station.isHidden boolValue] ) {
        [station.managedObjectContext performBlockAndWait:^{
            station.isHidden = @NO;
            [station.managedObjectContext save:nil];
        }];
    }

    RHCViewController *controller = [(UINavigationController *)self.window.rootViewController viewControllers].firstObject;
    if ( self.window.rootViewController.presentedViewController ) {
        [self.window.rootViewController dismissViewControllerAnimated:NO completion:^{
            [controller scrollToStation:station];
        }];
    } else {
        [controller scrollToStation:station];
    }

    return YES;
}


@end

File: VindsidenApp/RHCAppDelegate.swift
//
//  RHCAppDelegate.swift
//  VindsidenApp
//
//  Created by Ragnar Henriksen on 16/10/14.
//  Copyright (c) 2014 RHC. All rights reserved.
//

import UIKit
import VindsidenKit
import WatchConnectivity
import CoreSpotlight
import WidgetKit
import OSLog

@UIApplicationMain
class RHCAppDelegate: UIResponder, UIApplicationDelegate, WCSessionDelegate {

    var window: UIWindow?
    var connectionSession: WCSession?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

        self.window?.tintColor = UIColor.vindsidenGloablTintColor()

        if AppConfig.sharedConfiguration.applicationUserDefaults.integer(forKey: "selectedUnit") == 0 {
            AppConfig.sharedConfiguration.applicationUserDefaults.set(SpeedConvertion.toMetersPerSecond.rawValue, forKey: "selectedUnit")
            AppConfig.sharedConfiguration.applicationUserDefaults.synchronize()
        }

        if WCSession.isSupported() {
            let connectionSession = WCSession.default

            //if connectionSession.paired && connectionSession.watchAppInstalled {
                connectionSession.delegate = self
                connectionSession.activate()
            //}
        }

        DataManager.shared.cleanupPlots {
            WindManager.sharedManager.refreshInterval = 60
            WindManager.sharedManager.startUpdating()
        }

        application.setMinimumBackgroundFetchInterval(UIApplication.backgroundFetchIntervalMinimum)

        var performAdditionalHandling = true

        if let shortcutItem = launchOptions?[UIApplication.LaunchOptionsKey.shortcutItem] as? UIApplicationShortcutItem, let rootViewController = primaryViewController() {
            let didHandleShortcutItem = ShortcutItemHandler.handle(shortcutItem, with: rootViewController)
            performAdditionalHandling = !didHandleShortcutItem
        }

        ShortcutItemHandler.updateDynamicShortcutItems(for: application)

        return performAdditionalHandling
    }


    func application(_ application: UIApplication, willFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

        if let options = launchOptions {
            if let url = options[UIApplication.LaunchOptionsKey.url] as? URL {
                if let _ = url.host?.range(of: "station", options: .caseInsensitive) {
                    return true
                } else {
                    return false
                }
            }
        }

        return true
    }


    func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {

        if let _ = url.host?.range(of: "station", options: .caseInsensitive) {
            return openLaunchOptionsURL(url)
        }

        return false
    }


    func applicationWillResignActive(_ application: UIApplication) {
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        AppConfig.sharedConfiguration.presentReviewControllerIfCriteriaIsMet()
    }


    func applicationDidBecomeActive(_ application: UIApplication) {
        if WCSession.isSupported() {
            let connectionSession = WCSession.default

//            if connectionSession.paired && connectionSession.watchAppInstalled {
            connectionSession.delegate = self
            connectionSession.activate()
//            }
        }
    }


    func applicationWillTerminate(_ application: UIApplication) {
        DataManager.shared.saveContext()
    }


    func application(_ application: UIApplication, performFetchWithCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        Task { @MainActor in
            await WindManager.sharedManager.fetch()
            WidgetCenter.shared.reloadTimelines(ofKind: "VindsidenWidget")
            completionHandler(.newData)
        }
    }

    func application(_ application: UIApplication, handleWatchKitExtensionRequest userInfo: [AnyHashable: Any]?, reply: @escaping ([AnyHashable: Any]?) -> Void) {
        let taskID = application.beginBackgroundTask(expirationHandler: {})

        reply(["result": "not_updated"])
        application.endBackgroundTask(taskID)
    }


    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
        guard let window = window else {
            return .allButUpsideDown
        }

        if  window.traitCollection.userInterfaceIdiom == .pad  {
            return .all
        }

        return .allButUpsideDown
    }


    func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
        var didHandleShortcutItem = false

        let nc = self.window?.rootViewController as? UINavigationController
        let controller = primaryViewController()

        if let _ = nc?.presentedViewController {
            nc?.dismiss(animated: false, completion: { () -> Void in
                didHandleShortcutItem = ShortcutItemHandler.handle(shortcutItem, with: controller!)
            })
        } else {
            didHandleShortcutItem = ShortcutItemHandler.handle(shortcutItem, with: controller!)
        }

        completionHandler(didHandleShortcutItem)
    }


    // MARK: -

    
    func openLaunchOptionsURL( _ url: URL) -> Bool {
        guard let ident = url.pathComponents.last else {
            return false
        }

        var station: CDStation?

        if let stationId = Int(ident) {
            do {
                station = try CDStation.existingStationWithId(stationId, inManagedObjectContext: DataManager.shared.viewContext())
            } catch {
            }
        } else {
            station = CDStation.searchForStationName(ident, inManagedObjectContext: DataManager.shared.viewContext())
        }

        if  let found = station {
            if let hidden = found.isHidden, hidden.boolValue == true {
                found.managedObjectContext?.performAndWait({ () -> Void in
                    found.isHidden = false
                    do {
                        try found.managedObjectContext?.save()
                    } catch {
                        Logger.debugging.debug("Save failed")
                    }
                })
            }
        } else {
            return false
        }


        let nc = self.window?.rootViewController as? UINavigationController
        let controller = primaryViewController()

        if let _ = nc?.presentedViewController {
            nc?.dismiss(animated: false, completion: { () -> Void in
                controller!.scroll(to: station!)
            })
        } else {
            controller!.scroll(to: station!)
        }

        return true
    }


    // MARK: - Restoration


    func application(_ application: UIApplication, shouldRestoreApplicationState coder: NSCoder) -> Bool {
        return true
    }


    func application(_ application: UIApplication, shouldSaveApplicationState coder: NSCoder) -> Bool {
        return true
    }


    // MARK: - WC Session


    func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
        Logger.debugging.debug("Session: \(session)")
    }


    func sessionDidBecomeInactive(_ session: WCSession) {
        Logger.debugging.debug("Session: \(session)")
    }


    func sessionDidDeactivate(_ session: WCSession) {
        Logger.debugging.debug("Session: \(session)")
    }


    func sessionWatchStateDidChange(_ session: WCSession) {
        Logger.debugging.debug("Session: \(session)")
    }


    func sessionReachabilityDidChange(_ session: WCSession) {
        Logger.debugging.debug("Session: \(session)")
    }

    func session(_ session: WCSession, didReceiveMessage message: [String : Any], replyHandler: @escaping ([String : Any]) -> Void) {
        Logger.debugging.debug("Session: \(message)")
        replyHandler(["result": "not_updated"])
    }


    // MARK: - NSUserActivity

    override func updateUserActivityState(_ activity: NSUserActivity) {
        Logger.debugging.debug("\(activity)")
    }

    func application(_ application: UIApplication, willContinueUserActivityWithType userActivityType: String) -> Bool {
        return true
    }


    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        Logger.debugging.debug("Activity: \(userActivity.activityType) - \(String(describing: userActivity.userInfo))")


        if userActivity.activityType == CSSearchableItemActionType, let userInfo = userActivity.userInfo, let urlString = userInfo[CSSearchableItemActivityIdentifier] as? String {
            Logger.debugging.debug("URL STRING: \(urlString)")
            let url = URL(string: urlString)
            let success = openLaunchOptionsURL(url!)

            return success
        } else if let userInfo = userActivity.userInfo, let urlString = userInfo["urlToActivate"] as? String {
            let url = URL(string: urlString)
            let success = openLaunchOptionsURL(url!)

            return success
        }

        return false
    }

    
    func application(_ application: UIApplication, didFailToContinueUserActivityWithType userActivityType: String, error: Error) {
        Logger.debugging.debug("did fail to continue: \(userActivityType), \(error)")
    }


    func primaryViewController() -> RHCViewController? {
        let nc = self.window?.rootViewController as? UINavigationController
        let controller = nc?.viewControllers.first as? RHCViewController
        return controller
    }


    @objc func updateShortcutItems() {
        ShortcutItemHandler.updateDynamicShortcutItems(for: UIApplication.shared)
    }
}
